File: Native/Android/LunarConsole/lunarConsole/src/main/java/spacemadness/com/lunarconsole/utils/ConcurrentModificationList.java
//
//  ConcurrentModificationList.java
//
//  Lunar Unity Mobile Console
//  https://github.com/SpaceMadness/lunar-unity-console
//
//  Copyright 2015-2020 Alex Lementuev, SpaceMadness.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//


package spacemadness.com.lunarconsole.utils;

import java.util.ArrayList;
import java.util.List;

public abstract class ConcurrentModificationList<E> {
    protected final List<E> data;
    protected boolean updating;
    protected int removedCount;

    public ConcurrentModificationList(int capacity) {
        data = new ArrayList<>(capacity);
    }

    public void add(E element) {
        data.add(element);
    }

    public boolean remove(E element) {
        int index = data.indexOf(element);
        if (index != -1) {
            if (updating) {
                data.set(index, null); // replace element with 'tombstone'
                ++removedCount;
            } else {
                data.remove(index);
            }

            return true;
        }

        return false;
    }

    public void clear() {
        if (updating) {
            int i = 0;
            while (i < data.size()) {
                data.set(i, null);
                ++removedCount;
                ++i;
            }
        } else {
            data.clear();
        }
    }

    public int size() {
        return data.size();
    }

    public boolean isEmpty() {
        return data.size() == 0;
    }
}

File: Native/Android/LunarConsoleNew/lunarCore/src/main/java/spacemadness/com/lunarconsole/utils/ConcurrentModificationList.kt
package spacemadness.com.lunarconsole.utils

class ConcurrentModificationList<E : Any>(initialCapacity: Int = 0) {
    protected val data = ArrayList<E?>(initialCapacity)
    protected var updating = false
    protected var removedCount = 0

    val size get() = data.size
    val isEmpty get() = data.isEmpty()

    fun add(element: E) {
        data.add(element)
    }

    fun remove(element: E): Boolean {
        val index = data.indexOf(element)
        if (index != -1) {
            if (updating) {
                data[index] = null // replace element with 'tombstone'
                ++removedCount
            } else {
                data.removeAt(index)
            }
            return true
        }
        return false
    }

    fun clear() {
        if (updating) {
            var i = 0
            while (i < data.size) {
                data[i] = null
                ++removedCount
                ++i
            }
        } else {
            data.clear()
        }
    }

    @Suppress("PROTECTED_CALL_FROM_PUBLIC_INLINE") // TODO: this should be addressed
    inline fun forEach(callback: (index: Int, element: E) -> Unit) {
        try {
            updating = true
            val size = data.size
            var i = 0
            while (i < size) {
                val e = data[i]
                if (e != null) {
                    callback(i, e)
                }
                ++i
            }
        } finally {
            updating = false

            // clean up 'tombstones'
            var i = data.size - 1
            while (removedCount > 0 && i >= 0) {
                if (data[i] == null) {
                    data.removeAt(i)
                    removedCount -= 1
                }
                --i
            }
        }
    }
}

fun <T : Any> concurrentListOf(vararg elements: T) = ConcurrentModificationList<T>().apply {
    for (element in elements) {
        add(element)
    }
}